### PR TITLE
fix: adds a special case in nvim_cmd for cmds that take a count arg, neovim#23121

### PR DESCRIPTION
**Pull request by joshzhaa**
This location in nvim_cmd seems like the most appropriate to handle this case. If it isn't please let me know.

### CODE REVIEW

**Comment by famiu** [comment]
I'm instead thinking if `count` should set `EX_EXTRA | EX_NOSPC` like `nargs=?` does, if it doesn't cause any backwards compatibility issues. @bfredl thoughts?

File: src/nvim/api/command.c [hunk 1]
```diff
@@ -458,7 +458,12 @@ String nvim_cmd(uint64_t channel_id, Dict(cmd) *cmd, Dict(cmd_opts) *opts, Error
       argc_valid = true;
       break;
     default:
-      argc_valid = args.size == 0;
+      // if the cmd can take a count argument, it optionally has 1 arg
+      if (ea.argt & EX_COUNT) {
+        argc_valid = args.size == 0 || args.size == 1;
+      } else {
+        argc_valid = args.size == 0;
+      }
```
Comment (famiu) [inline src/nvim/api/command.c:466]:
```suggestion
      // if the cmd can take a count argument, it optionally has 1 arg
      argc_valid = args.size == 0 || ((ea.argt & EX_COUNT) && args.size == 1);
```

**Comment by bfredl** [comment]
It is probably fine but I would like to see a test for a command which needs this.

**Comment by famiu** [comment]
> It is probably fine but I would like to see a test for a command which needs this.

I agree. So I assume we agree that modifying count to set `EX_EXTRA | EX_NOSPC` (if possible) is better than adding a special exception in `nvim_cmd`, just to be clear?

**Comment by bfredl** [comment]
let's try it and see if it breaks any existing test.

**Comment by zeertzjq** [comment]
> > It is probably fine but I would like to see a test for a command which needs this.
> 
> I agree. So I assume we agree that modifying count to set `EX_EXTRA | EX_NOSPC` (if possible) is better than adding a special exception in `nvim_cmd`, just to be clear?

That sounds like a hack. I think modifying `nvim_cmd` is better, although probably in a more general way.

**Comment by famiu** [comment]
> > > It is probably fine but I would like to see a test for a command which needs this.
> > 
> > 
> > I agree. So I assume we agree that modifying count to set `EX_EXTRA | EX_NOSPC` (if possible) is better than adding a special exception in `nvim_cmd`, just to be clear?
> 
> That sounds like a hack. I think modifying `nvim_cmd` is better, although probably in a more general way.

Interesting. Because to me, modifying `nvim_cmd` is what seems like a hack tbh. A count is treated as an argument and therefore should modify the argument flags, or at least that's my rationale behind it.

**Comment by zeertzjq** [comment]
Actually neither approaches work correctly.

This should give an error about argument count:
```vim
vim.cmd.undo('1x')
```
Neither of the two approaches mentioned in this PR give this error.

The only proper solution I can think of is moving the argument count check after the `parse_count()`.

**Comment by zeertzjq** [comment]
Another possible solution is moving the `parse_count()` in `execute_cmd()` to `nvim_cmd()`, before the check for argument count. This may also avoid some unexpected behavior (e.g. two arguments both parsed as count) with 'inccommand', but is a bit tricky to do.

**Pull request closed without merging**